router/oauthResponseHandler/oauthResponseHandler.go

```diff
@@ -0,0 +1,546 @@
+		disableDestStats.errorMessage = msg
+		disableDestStats.SendCountStat()
+		return http.StatusBadRequest, msg
+	}
+
+	authErrHandler.logger.Debugf("[%s request] :: (Write) Disable Response received : %s\n", loggerNm, respBody)
+	disableDestStats.statName = "disable_destination_success"
+	disableDestStats.errorMessage = ""
+	disableDestStats.SendCountStat()
+
+	return statusCode, fmt.Sprintf(`{response: {isDisabled: %v, activeRequest: %v}`, !disableDestRes.Enabled, false)
+}
+
+func processResponse(resp *http.Response) (statusCode int, respBody string) {
+	var respData []byte
+	var ioUtilReadErr error
+	if resp != nil && resp.Body != nil {
+		respData, ioUtilReadErr = ioutil.ReadAll(resp.Body)
+		defer resp.Body.Close()
+		if ioUtilReadErr != nil {
+			return http.StatusInternalServerError, ioUtilReadErr.Error()
+		}
+	}
+	//Detecting content type of the respData
+	contentTypeHeader := http.DetectContentType(respData)
+	//If content type is not of type "*text*", overriding it with empty string
+	if !(strings.Contains(strings.ToLower(contentTypeHeader), "text") ||
+		strings.Contains(strings.ToLower(contentTypeHeader), "application/json") ||
+		strings.Contains(strings.ToLower(contentTypeHeader), "application/xml")) {
+		respData = []byte("")
+	}
+
+	return resp.StatusCode, string(respData)
+}
+
+func (authErrHandler *OAuthErrResHandler) cpApiCall(cpReq *ControlPlaneRequestT) (int, string) {
+	var reqBody *bytes.Buffer
+	var req *http.Request
+	var err error
+	if router_utils.IsNotEmptyString(cpReq.Body) {
+		reqBody = bytes.NewBufferString(cpReq.Body)
+		req, err = http.NewRequest(cpReq.Method, cpReq.Url, reqBody)
+	} else {
+		req, err = http.NewRequest(cpReq.Method, cpReq.Url, nil)
+	}
+	if err != nil {
+		authErrHandler.logger.Errorf("[%s request] :: destination request failed: %+v\n", loggerNm, err)
+		// Abort on receiving an error in request formation
+		return http.StatusBadRequest, err.Error()
+	}
+	// Authorisation setting
+	req.SetBasicAuth(workspaceToken, "")
+
+	// Set content-type in order to send the body in request correctly
+	if router_utils.IsNotEmptyString(cpReq.ContentType) {
+		req.Header.Set("Content-Type", cpReq.ContentType)
+	}
+
+	authErrHandlerTimeStart := time.Now()
+	res, doErr := authErrHandler.client.Do(req)
+	stats.NewTaggedStat("cp_request_latency", stats.TimerType, stats.Tags{
+		"url":         cpReq.Url,
+		"destination": cpReq.destName,
+	}).SendTiming(time.Since(authErrHandlerTimeStart))
+	authErrHandler.logger.Debugf("[%s request] :: destination request sent\n", loggerNm)
+	if res.Body != nil {
+		defer res.Body.Close()
+	}
+	if doErr != nil {
+		// Abort on receiving an error
+		authErrHandler.logger.Errorf("[%s request] :: destination request failed: %+v\n", loggerNm, doErr)
+		return http.StatusBadRequest, doErr.Error()
+	}
+	statusCode, resp := processResponse(res)
+	return statusCode, resp
+}
+
+func (resHandler *OAuthErrResHandler) NewMutex(id string, errCategory string) error {
```

Comment:
A few concerns regarding this method:
1. It seems to me that this method should not be exported, thus 
```suggestion
func (resHandler *OAuthErrResHandler) newMutex(id string, errCategory string) error {
``` (lower case)
2. Is this method is control-coupled by the `errCategory` function

A suggestion here would be to do something like this:
```suggestion
 setMapMutex(map[string]*sync.RWMutex, id string)
 ```
 
 This way you can avoid:
 ```
 switch errCategory {
	case DISABLE_DEST:
		mutexMap = resHandler.destLockMap
	case REFRESH_TOKEN:
		mutexMap = resHandler.accountLockMap
	default:
		// when errorCategory i empty
		resHandler.logger.Debugf("[%s request] :: Case missing for mutex for %s\n", loggerNm, id)
		return fmt.Errorf(`except %v, %v error category is not supported`, DISABLE_DEST, REFRESH_TOKEN)
	}
```

Also makes the function, generic enough to be used in other places as well.

3. If you care about performance, consider using a Read lock for checking if the id already exists and then switch to a write `Lock` .

Reply:
```golang
setMapMutex(map[string]*sync.RWMutex, id string)
```
Sorry couldn't completely understand your perspective about this. Can you explain it a bit more ?